Various Enhancements for the `testnet` Binary

jacderida opened the pull request:
- 1f49a23 **feat: remove network contacts from `testnet` bin**

  While we do still have the concept of contacting other peers--via the `--peer` argument or
  `SAFE_PEERS` environment variable--we are not using a 'network contacts' file any more, so we don't
  need this argument for the `testnet` binary.

- c5f20ba **refactor: obtain genesis peer id directly**

  BREAKING CHANGE: Remove the `verify-nodes` feature in favour of just including this code in the
  binary. It only increases the size by a few megabytes and the binary is much more useful with this
  code.

  Refactor the genesis launching process to use the gRPC generated code to obtain the genesis peer ID,
  rather than using the `safe_rpc_client` example. The latter was actually built during the process,
  which involved having the source code checked out. This is probably an assumption we shouldn't make
  of a released binary. It's intended to be able to work stand alone, without the source.

- f779592 **tests: restore sn_testnet unit tests**

  The `sn_testnet` crate was more or less just copy/pasted from the previous safe_network repo and
  its behaviour was quickly changed for dealing with the new `safenode` binary. For quickness, the
  unit tests were subverted.

  These unit tests are now brought into line with the way the new `safenode` process operates and we
  can now keep them running for preventing regressions when things change.

- 5cd6409 **feat: remove node directory management**

  The testnet binary is updated to remove any node directory management, which is now performed by the
  `safenode` process.

  This simplifies both the code and test cases, and also removes a few cases that are no longer
  necessary.

- a1eb03e **feat: provide a `--clean` flag**

  Provides a `--clean` flag to remove all previous node root directories under the data directory,
  which is platform specific.

- 24dd50a **ci: remove use of `verify-nodes` feature**

  This feature has now been removed from the `testnet` binary. The functionality is now in the binary
  by default.

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 04 Jul 23 22:14 UTC
This pull request includes the following changes:

1. The file diff removes the initialization of logging targets and the `init_logging` function call. It also removes the import of the `Level` enum from `tracing_core`.

2. The file diff for `main.rs` includes changes such as adding environment variables, modifying commands for starting a local network, updating benchmarks section, modifying benchmark names and parameters, and updating commands for checking memory usage.

3. The file `replication.rs` has changes related to adjusting the replication behavior of the `Node` implementation.

4. The file `sequential_transfers.rs` has changes related to logging configuration and do not affect the core functionality.

5. The `Cargo.toml` file has changes related to dependencies, versions, and package updates.

6. The changes in `mod.rs` include importing additional modules, initializing logging configuration, and removing logging related code.

7. The changes in `check_testnet.rs` include updating log output and console messages, new function addition, and modifying log file checking.

8. The file `build.rs` has changes related to removing unused import and modifying the compilation process of `safenode.proto`.

9. The file `data_with_churn.rs` has changes related to logging and removing records from specific directories.

10. The file `cli.rs` has changes related to importing additional modules and adding new options and subcommands.

11. The file `.github/workflows/nightly.yml` has changes related to modifying commands, removing trailing newline, and updating run flags.

12. The file `sn_logging.rs` has changes related to adding enums, implementing methods, and adding a new function for initializing logging configuration.

13. The file `safenode_rpc_client.rs` has changes related to logging initialization and imports.

14. The diff of the file `data_with_churn.rs` includes changes related to importing additional modules, changing arguments for `init_logging` function, replacing path with variable, and removing a `println!` statement.

15. The changes in `cli.rs` include importing additional modules, adding a function for parsing log output, and adding a struct with new fields.

16. The file `memcheck.yml` has changes related to setting environment variables, adding steps for package installation, modifying commands for building and starting nodes, modifying commands for checking network and replication, modifying commands for checking client files, modifying commands for monitoring memory usage, and modifying commands for creating log files archive.

17. The diff in `build.rs` removes the conditional compilation feature and changes the location of the `tonic_build::compile_protos` line.

Please review these changes and let me know if you need any further clarification.
<!-- reviewpad:summarize:end -->

joshuef commented:
Would be good to get an addition to this for readme, otherwise LGTM, lets get it in and give it a whirl

lionel1704 commented:
Will re-open this later.

jacderida committed:
feat: remove network contacts from `testnet` bin

While we do still have the concept of contacting other peers--via the `--peer` argument or
`SAFE_PEERS` environment variable--we are not using a 'network contacts' file any more, so we don't
need this argument for the `testnet` binary.

jacderida committed:
refactor: obtain genesis peer id directly

BREAKING CHANGE: Remove the `verify-nodes` feature in favour of just including this code in the
binary. It only increases the size by a few megabytes and the binary is much more useful with this
code.

Refactor the genesis launching process to use the gRPC generated code to obtain the genesis peer ID,
rather than using the `safe_rpc_client` example. The latter was actually built during the process,
which involved having the source code checked out. This is probably an assumption we shouldn't make
of a released binary. It's intended to be able to work stand alone, without the source.

jacderida committed:
tests: restore sn_testnet unit tests

The `sn_testnet` crate was more or less just copy/pasted from the previous safe_network repo and
its behaviour was quickly changed for dealing with the new `safenode` binary. For quickness, the
unit tests were subverted.

These unit tests are now brought into line with the way the new `safenode` process operates and we
can now keep them running for preventing regressions when things change.

jacderida committed:
feat: provide `--log-output-dest` arg for `safenode`

BREAKING CHANGE: the `--log-dir` argument has been removed in favour of `--log-output-dest`.

This argument uses a 'special' value, which can be "stdout", "default", or a custom directory path.
The "default" is the default root directory, plus "logs" appended to that path.

The `sn_logging` crate now provides a `LogOutputDest` enum which makes for a nicer interface on the
logging initialisation, removing the need for an `Option<PathBuf>` to specify output to a file.

jacderida committed:
feat: provide `--log-output-dest` arg for `safenode`

BREAKING CHANGE: the `--log-dir` argument has been removed in favour of `--log-output-dest`.

This argument uses a 'special' value, which can be "stdout", "default", or a custom directory path.
The "default" is the default root directory, plus "logs" appended to that path.

The `sn_logging` crate now provides a `LogOutputDest` enum which makes for a nicer interface on the
logging initialisation, removing the need for an `Option<PathBuf>` to specify output to a file.

jacderida committed:
feat: introduce `--log-format` arguments

BREAKING CHANGE: for both `safe` and `safenode` the `--json-log-output` argument is removed in
favour of this new `--log-format` argument.

The argument functions in a similar fashion to `--log-output-dest`, in that it uses a 'special'
value, which for the time being can either be "default" or "json". It has the potential to be
expanded.

jacderida committed:
ci: update node logging paths

In light of the changes to the logging setup, the testnet tool is modified such that it will use the
default root and logging directories.

The merge workflow is also updated to look for logs in the new directories. For the correct
usage of paths, in some cases this meant using different mechanisms on Windows vs. Unix.

The memcheck workflow was also updated to the same effect, but it only runs on Linux, so
cross-platform considerations weren't required.

joshuef committed:
feat(node): make RPC::Restart wipe node data entirely

joshuef committed:
chore(node): re add build_info to node

joshuef committed:
fix(node): get PeerId on every node startup loop

jacderida committed:
feat: remove node directory management

The testnet binary is updated to remove any node directory management, which is now performed by the
`safenode` process.

This simplifies both the code and test cases, and also removes a few cases that are no longer
necessary.

jacderida committed:
feat: provide a `--clean` flag

Provides a `--clean` flag to remove all previous node root directories under the data directory,
which is platform specific.

jacderida committed:
ci: remove use of `verify-nodes` feature

This feature has now been removed from the `testnet` binary. The functionality is now in the binary
by default.

jacderida committed:
fix: resolve post-rebase compilation issues

A difficult rebase left the `sn_node/src/main.rs` file with content from both sides of the merge,
which meant it wouldn't compile properly.

jacderida commented:
I'm going to cherry pick these commits onto a new branch.
